src/Data/Array/Mutable/Linear.hs

```diff
@@ -11,75 +15,123 @@
+-- Note that this does throw an exception and that there's a reasonable
+-- coherent view and method to using exceptions in Haskell!
+-- See https://www.tweag.io/posts/2020-04-16-exceptions-in-haskell.html.
+alloc :: HasCallStack =>
+  Int -> a -> (Array a #-> Unrestricted b) -> Unrestricted b
+alloc size x f
+  | size > 0 = f (Array size (Unsafe.newMutArr size x))
+  | otherwise = error $ "Trying to allocate an array of size " ++ show size
+
+fromList :: HasCallStack =>
+  [a] -> (Array a #-> Unrestricted b) -> Unrestricted b
+fromList [] _ = error $ "Trying to allocate from an empty list."
+fromList list@(x:_) (f :: Array a #-> Unrestricted b) =
+  alloc (Prelude.length list) x insertThenf
+  where
+    insertThenf :: Array a #-> Unrestricted b
+    insertThenf arr = f (doWrites (zip list [0..]) arr)
+
+    doWrites :: [(a,Int)] -> Array a #-> Array a
+    doWrites [] arr = arr
+    doWrites ((a,ix):xs) arr = doWrites xs (write arr ix a)
 
 -- # Mutations and Reads
-----------------------------
+-------------------------------------------------------------------------------
 
-length :: Array a #-> (Int, Array a)
-length = Linear.toLinear unsafeLength
+length :: Array a #-> (Array a, Int)
+length = Unsafe.toLinear unsafeLength
   where
-    unsafeLength :: Array a -> (Int, Array a)
-    unsafeLength v@(Array size _) = (size, v)
+    unsafeLength :: Array a -> (Array a, Int)
+    unsafeLength v@(Array size _) = (v, size)
 
 write :: HasCallStack => Array a #-> Int -> a -> Array a
-write = Linear.toLinear writeUnsafe
+write = Unsafe.toLinear writeUnsafe
   where
     writeUnsafe :: Array a -> Int -> a -> Array a
     writeUnsafe arr@(Array size mutArr) ix val
       | indexInRange size ix =
-        case writeMutArr mutArr ix val of
+        case Unsafe.writeMutArr mutArr ix val of
           () -> arr
       | otherwise = error "Write index out of bounds."
 
 read :: HasCallStack => Array a #-> Int -> (Array a, a)
-read = Linear.toLinear readUnsafe
+read = Unsafe.toLinear readUnsafe
   where
     readUnsafe :: Array a -> Int -> (Array a, a)
     readUnsafe arr@(Array size mutArr) ix
-      | indexInRange size ix = (arr, readMutArr mutArr ix)
+      | indexInRange size ix = (arr, Unsafe.readMutArr mutArr ix)
       | otherwise = error "Read index out of bounds."
 
+-- | Using first element as seed, resize to a constant array
 resize :: HasCallStack => Int -> Array a #-> Array a
 resize newSize (Array _ mutArr) =
-  Array newSize (resizeMutArr mutArr newSize)
+  Array newSize (Unsafe.newMutArr newSize (Unsafe.readMutArr mutArr 0))
 
--- Resize with a seed value for the newly allocated elements
+-- | Resize to a new constant array given a seed value
 resizeSeed :: HasCallStack => Int -> a -> Array a #-> Array a
-resizeSeed newSize seed (Array _ mutArr) =
-  Array newSize (resizeMutArrSeed mutArr seed newSize)
+resizeSeed newSize seed (Array _ _) =
+  Array newSize (Unsafe.newMutArr newSize seed)
+
+-- XXX: Replace with toVec
```

Comment:
I mean, it's a fine function to have. Having an `O(1)` freeze would be nice too, of course.